### PR TITLE
[mlir][linalg] Vectorize unpack op without masking

### PR DESCRIPTION
**Pull request by pashu123**
Enables vectorization of unpack op in the case of unknown vector size.
The vector sizes are determined by the result's shape.

### CODE REVIEW

**Comment by pashu123** [comment]
@hanhanW I am in the process of writing tests. Thanks.

**Comment by pashu123** [comment]
> [nit] Could you trim the commit subject? It's much easier to read if it fits in one line. Also:
> 
> * https://mlir.llvm.org/getting_started/Contributing/#commit-messages --> https://cbea.ms/git-commit/#limit-50
> 
> Thanks :)

Thanks for the reference. I have updated the message and the body. PTAL.

File: mlir/test/Dialect/Linalg/vectorization.mlir
Comment (hanhanW) [inline mlir/test/Dialect/Linalg/vectorization.mlir:1]: We need a test for unpack which also slices output. E.g.,

```
%0 = tensor.unpack %source
  inner_dims_pos = [0, 1]
  inner_tiles = [32, 16]
  into %dest : tensor<8x8x32x16xf32> -> tensor<255x127xf32>
```

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (banach-space) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1608]: Similar comment as for https://github.com/llvm/llvm-project/pull/89119 ...

IMHO it would make more sense to introduce `useInBoundsInsteadOfMasking` in place of `doMasking`. 

The longer name better captures the underlying logic ("replace masks with in_bounds attr"). Indeed, the actual meaning of the bool is much narrower than `doMasking`.

Naming is hard and I am being a bit pedantic. And also a bit mindful of our future selves when reading this in 6 months :)

File: mlir/test/Dialect/Linalg/vectorization.mlir
Comment (pashu123) [reply mlir/test/Dialect/Linalg/vectorization.mlir:1]: Would the vector sizes for this case be `inner_tiles[x] * source_dim[inner_dims_pos[x]] for x in len(inner_tiles)` and then the inbounds will be set accordingly?

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (hanhanW) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1625]: nit: We don't need to cast it SmallVector to ArrayRef... so maybe replace it with `readMaksShape`

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (hanhanW) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1773]: I think we don't need the `if` here.. so maybe just put the conditions together in the initialization. E.g.,

```
bool satisfyEmptyCond = inputVectorSizes.empty() && .. &&..;
```

File: mlir/test/Dialect/Linalg/vectorization.mlir
Comment (hanhanW) [reply mlir/test/Dialect/Linalg/vectorization.mlir:1]: Yes, I think so. The inbounds (of the xfer_write op) will be set accordingly.

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (hanhanW) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1590]: two nit:

1. Move this into the if statement because it is only used there.
2. Avoid brace-list initialization, instead we can use constructor directly.

```suggestion
  SmallVector<int64_t> initVectorShape(sourceShape.take_front(destSize));
```

We prefer using `()` constructor or `=` assignment instead of brace-list initialization. This kind of code is dangerous when the type is not an integer type. E.g.,

```
std::vector<std::string> strings{2}; // A vector of two empty strings.
std::vector<int> ints{2};            // A vector containing only the integer 2.
```

See https://abseil.io/tips/88 for more details.

**Review by hanhanW — APPROVED** [review]
thanks!

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (pashu123) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp]: 1. This Can't be done since inputVectorSizes references initVectorShape, and hence, it has to live outside of the if scope.

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (banach-space) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1599]: Is yet another variable, `initVectorShape`, really needed? Also, see my comment below.

```suggestion
  if (inputVectorSizes.empty()) {
    inputVectorSizes.resize(sourceShape.take_front(destSize))
    if (!outerDimsPerm.empty())
      applyPermutationToVector(inputVectorSizes, outerDimsPerm);
    for (auto [i, pos] : llvm::enumerate(innerDimPos))
      inputVectorSizes[pos] *= innerTiles[i];

    useInBoundsInsteadOfMasking = true;
  }
```

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (banach-space) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1602]: At this point `useInBoundsInsteadOfMasking` is already set - why do we bother defining and calculating `readMaskShape` if `useInBoundsInsteadOfMasking` is true? It feels like something that should be factored out to a dedicated hook, e.g. `computeReadMaskShapeForUnpackOp`, and then:
```cpp
SmallVector<int64_t> readMaskShape;
if (!useInBoundsInsteadOfMasking)
  readMaskShape = computeReadMaskShapeForUnpackOp();  
```

Also, it looks like the shape calculation for `readMaskShape` and `initVectorShape` are duplicated? Again, why not introduce a dedicated hook for that?

File: mlir/test/Dialect/Linalg/vectorization.mlir
Comment (banach-space) [inline mlir/test/Dialect/Linalg/vectorization.mlir:1]: Could you add `in_bounds` to these tests?

**Review by banach-space — CHANGES_REQUESTED** [review]
Please could you address my comments before landing this? It feels that there's scope for better code re-use.

File: mlir/test/Dialect/Linalg/vectorization.mlir
Comment (hanhanW) [reply mlir/test/Dialect/Linalg/vectorization.mlir:1]: I was thinking the same thing, but it is doable in the current tests. They are hidden because they are all false. So maybe @pashu123 can add a test similar to below, then we can test it.


```
// the output is 64x127, not 63x127. The inbounds would be (true, false), and they should present in the IR.
%0 = tensor.unpack %source outer_dims_perm = [1, 0] inner_dims_pos = [0, 1] inner_tiles = [16, 16] into %dest : tensor<8x4x16x16xf32> -> tensor<64x127xf32>
```

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (pashu123) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1599]: We won't be able to modify/resize the `inputVectorShape` var since it's an `ArrayRef`, but we can point it to another variable. In that sense, it's needed.

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (pashu123) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1602]: Well, I think `readMaskShape` is not named properly. It's actually `readVectorSizes` and we have to unconditionally define it.

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (hanhanW) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1600]: I think we can't do the trick because it is `ArrayRef` type. We can only apply permutation on `SmallVector`.

Comment (pashu123) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp]: I've tried to simplify it so we don't have the `else` block. Also, we can't use `vectorSizes = inputVectorSizes` SmallVectors can't point to ArrayRefs but the other way around is possible.

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (pashu123) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp]: Yes, we need to calculate it once. I've renamed this to `readVectorSizes` since the mask is generated by createReadOrMaskedRead based on the vectorSizes. The same goes for `writeMaskShape`.

**Review by banach-space — APPROVED** [review]
LGTM, modulo a few minor suggestions, thanks!

This is looking much better now and is much easier to follow, thank you @pashu123 !

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (pashu123) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1592]: 1. There's actually a check performed here: https://github.com/llvm/llvm-project/blob/64013f94eaea2f0e1bf752cc1ffec3989f07d358/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1676. Only if the destination type is static can we use vectorSizes; otherwise, we resort to something else.
2. check is performed here: https://github.com/llvm/llvm-project/blob/64013f94eaea2f0e1bf752cc1ffec3989f07d358/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1798

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (pashu123) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1591]: Done. Thanks. I need to improve my comments.

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp
Comment (banach-space) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1592]: Yeah, looks like this condition is indeed checked in 2. above, thanks! 

That's a "pre-condition" though - no harm in adding an additional assert to document assumptions made in this method.

In any case, it's just a nice to have :)

Comment (pashu123) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1592]: Added. thanks.

Comment (banach-space) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:1596]: Thanks again for working on this - that's greatly appreciated 🙏🏻

**Pull request merged**
